controllers/upgrade_controller.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return fmt.Errorf("failed to convert channel name %q to semver %v", channel, err)
+	}
+
+	// decide whether to patch the subscription or not
+	platformVersion, _ := semver.Make(r.PlatformVersion)
+	if desiredVersion.Major > platformVersion.Major || desiredVersion.Minor > platformVersion.Minor {
+		r.log.Info("backing off from updating subscription as operator version can become ahead of platform version")
+	} else if desiredVersion.Major == channelVersion.Major && desiredVersion.Minor == channelVersion.Minor+1 {
+		// initiate auto upgrade by patching the channel name of the ocsClientOperatorSubscription
+		patchInfo := []struct {
+			Op    string `json:"op"`
+			Path  string `json:"path"`
+			Value string `json:"value"`
+		}{
+			{
+				Op:    "replace",
+				Path:  "/spec/channel",
+				Value: getChannelFromVersion(desiredVersion),
+			},
+		}
+		jsonPatch, _ := json.Marshal(patchInfo)
+		if err = r.patch(ocsClientOperatorSubscription, client.RawPatch(types.JSONPatchType, jsonPatch)); err != nil {
+			return fmt.Errorf("failed to update subscription %q/%q channel name to %q. %v",
+				ocsClientOperatorSubscription.Namespace, ocsClientOperatorSubscription.Name, ocsClientOperatorSubscription.Spec.Channel, err)
+		}
+		r.log.Info("updated subscription channel name",
+			"namespace", ocsClientOperatorSubscription.Namespace, "name", ocsClientOperatorSubscription.Name,
+			"to", ocsClientOperatorSubscription.Spec.Channel)
+	}
+
+	// decide whether to surface we are upgradeable or not, subscription should not be ahead of both desiredVersion and platformVersion
+	if desiredVersion.Major < channelVersion.Major || desiredVersion.Minor < channelVersion.Minor {
+		// someone manually might have initiated an upgrade, best effort to stop it via conditions
+		return fmt.Errorf("subscription channel name is changed outside of operator  (subscription is ahead of desiredVersion)")
+	}
+
+	if platformVersion.Major < channelVersion.Major || platformVersion.Minor < channelVersion.Minor {
+		return fmt.Errorf("subscription channel name is changed outside of operator (subscription is ahead of platform)")
+	}
+	// TODO (lgangava): as of now there's no way to distinguish desiredVersion vs stale desiredVersion
+	// 	and we aren't checking dependents for updating operatorcondition
+
+	return nil
+}
+
+func (r *UpgradeReconciler) setOperatorCondition(isUpgradeable metav1.ConditionStatus, message string) error {
+	return r.OperatorCondition.
+		Set(r.ctx, isUpgradeable, conditions.WithReason(upgradeConditionReason), conditions.WithMessage(message))
+}
+
+func (r *UpgradeReconciler) list(list client.ObjectList, opts ...client.ListOption) error {
+	return r.Client.List(r.ctx, list, opts...)
+}
+
+func (r *UpgradeReconciler) patch(obj client.Object, patch client.Patch, opts ...client.PatchOption) error {
+	return r.Client.Patch(r.ctx, obj, patch, opts...)
+}
+
+func getOldestVersion(versions []string) (semver.Version, error) {
+	if len(versions) < 1 {
```

Comment:
Missing a preceding nil check (which could have been more straightforward without using `make` upon init)

Reply:
nope, `len` is an all rounder, it's always better to use that for checking emptiness on items
``` go
> var first []string
: 0xc00011c0e0
> var second []string = nil
: 0xc0004040f0
> third := []string{}
: []
> fourth := make([]string, 0, 5)
: []
> fifth := make([]string, 5)
: [    ]
> fmt.Println(len(first)==0)
true
: 5
> fmt.Println(len(second)==0)
true
: 5
> fmt.Println(len(third)==0)
true
: 5
> fmt.Println(len(fourth)==0)
true
: 5
> fmt.Println(len(fifth)==0)
false
: 6
```